File: basic_dsa_two_hundred_two.java
/**
 * Basic DSA Problems - Product of Array Except Self
 * Return array where output[i] = product of all elements except nums[i] (no division, O(n))
 */
public class basic_dsa_two_hundred_two {

    // Problem: Return array such that out[i] = product of all nums except nums[i]. O(n), no division.

    /**
     * Two passes: first store prefix products in output, then multiply by suffix from the right.
     * Time O(n), Space O(1) excluding output
     */
    public static int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] out = new int[n];
        out[0] = 1;
        for (int i = 1; i < n; i++)
            out[i] = out[i - 1] * nums[i - 1];
        int suffix = 1;
        for (int i = n - 1; i >= 0; i--) {
            out[i] *= suffix;
            suffix *= nums[i];
        }
        return out;
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Product of Array Except Self");
        System.out.println("=".repeat(60));

        int[] nums = {1, 2, 3, 4};
        System.out.println("Result -> " + java.util.Arrays.toString(productExceptSelf(nums)));
    }
}

File: basic_dsa_two_hundred_two.py
"""
Basic DSA - Product of Array Except Self
Given an integer array nums, return an array answer such that answer[i] is the
product of all the elements of nums except nums[i].
You must write an algorithm that runs in O(n) time and without using division.
"""


def product_except_self(nums):
    """
    Prefix products then multiply by suffix products in a second pass.
    Time Complexity: O(n)
    Space Complexity: O(1) extra (output excluded)
    """
    n = len(nums)
    out = [1] * n

    pref = 1
    for i in range(n):
        out[i] = pref
        pref *= nums[i]

    suff = 1
    for i in range(n - 1, -1, -1):
        out[i] *= suff
        suff *= nums[i]

    return out


if __name__ == "__main__":
    print("=" * 60)
    print("Problem: Product of Array Except Self")
    print("=" * 60)

    print(product_except_self([1, 2, 3, 4]))  # [24, 12, 8, 6]
    print(product_except_self([-1, 1, 0, -3, 3]))  # [0, 0, 9, 0, 0]
